rmw_zenoh_cpp/src/rmw_service.cpp

```diff
@@ -0,0 +1,419 @@
+
+  // ASSERTIONS ================================================================
+  RMW_CHECK_ARGUMENT_FOR_NULL(node, nullptr);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    node,
+    node->implementation_identifier,
+    eclipse_zenoh_identifier,
+    return nullptr
+  );
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(service_name, nullptr);
+  if (strlen(service_name) == 0) {
+    RMW_SET_ERROR_MSG("service name is empty string");
+    return nullptr;
+  }
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(qos_profile, nullptr);
+  RMW_CHECK_ARGUMENT_FOR_NULL(type_supports, nullptr);
+
+  // OBTAIN ALLOCATOR ==========================================================
+  rcutils_allocator_t * allocator = &node->context->options.allocator;
+
+  // VALIDATE SERVICE NAME =====================================================
+  int * validation_result = static_cast<int *>(allocator->allocate(sizeof(int), allocator->state));
+
+  rmw_validate_full_topic_name(service_name, validation_result, nullptr);
+
+  if (*validation_result == RMW_TOPIC_VALID
+      || qos_profile->avoid_ros_namespace_conventions) {
+    allocator->deallocate(validation_result, allocator->state);
+  } else {
+    RMW_SET_ERROR_MSG("service name is malformed!");
+    allocator->deallocate(validation_result, allocator->state);
+    return nullptr;
+  }
+
+  // OBTAIN TYPESUPPORT ========================================================
+  const rosidl_service_type_support_t * type_support = get_service_typesupport_handle(
+    type_supports, RMW_ZENOH_CPP_TYPESUPPORT_C
+  );
+
+  if (!type_support) {
+    type_support = get_service_typesupport_handle(type_supports, RMW_ZENOH_CPP_TYPESUPPORT_CPP);
+    if (!type_support) {
+      RCUTILS_LOG_INFO("%s", service_name);
+      RMW_SET_ERROR_MSG("type support not from this implementation");
+      return nullptr;
+    }
+  }
+
+  // CREATE SERVICE ============================================================
+  rmw_service_t * service = static_cast<rmw_service_t *>(
+    allocator->allocate(sizeof(rmw_service_t), allocator->state)
+  );
+  if (!service) {
+    RMW_SET_ERROR_MSG("failed to allocate rmw_service_t");
+    allocator->deallocate(service, allocator->state);
+    return nullptr;
+  }
+
+  // Populate common members
+  service->implementation_identifier = eclipse_zenoh_identifier;  // const char * assignment
+
+  service->service_name = rcutils_strdup(service_name, *allocator);  // const char * assignment
+  if (!service->service_name) {
+    RMW_SET_ERROR_MSG("failed to allocate service name");
+    allocator->deallocate(service, allocator->state);
+    return nullptr;
+  }
+
+  service->data = static_cast<rmw_service_data_t *>(
+    allocator->allocate(sizeof(rmw_service_data_t), allocator->state)
+  );
+  if (!service->data) {
+    RMW_SET_ERROR_MSG("failed to allocate service data");
+    allocator->deallocate(service->data, allocator->state);
+    allocator->deallocate(service, allocator->state);
```

Comment:
Does this leak `service->service_name`?

Reply:
Yep...

Updated: 23497d4430f890dc62fb9995c66fb2be8602b51e

(Sorry for the massive commit, it's kind of intertwined...)